File: neetcode-150/sliding-window/longest-substring-without-repeating-characters/solution.js
/**
 * Time Complexity: O(n) - Each character visited at most twice
 * Space Complexity: O(min(m,n)) - Where m is charset size
 */
var lengthOfLongestSubstring = function(s) {
    if (s.length === 0) return 0;
    
    const charCount = {};
    let left = 0;
    let maxLength = 0;
    
    for (let right = 0; right < s.length; right++) {
        const rightChar = s[right];
        charCount[rightChar] = (charCount[rightChar] || 0) + 1;
        
        // Contract window if duplicates found
        while (charCount[rightChar] > 1) {
            const leftChar = s[left];
            charCount[leftChar]--;
            left++;
        }
        
        maxLength = Math.max(maxLength, right - left + 1);
    }
    
    return maxLength;
};

// Alternative approach using Set
var lengthOfLongestSubstringSet = function(s) {
    if (s.length === 0) return 0;
    
    const charSet = new Set();
    let left = 0;
    let maxLength = 0;
    
    for (let right = 0; right < s.length; right++) {
        const rightChar = s[right];
        
        // Contract window if duplicate found
        while (charSet.has(rightChar)) {
            charSet.delete(s[left]);
            left++;
        }
        
        charSet.add(rightChar);
        maxLength = Math.max(maxLength, right - left + 1);
    }
    
    return maxLength;
};

// Optimized version with character position tracking
var lengthOfLongestSubstringOptimized = function(s) {
    const charIndex = {};
    let left = 0;
    let maxLength = 0;
    
    for (let right = 0; right < s.length; right++) {
        const rightChar = s[right];
        
        if (charIndex[rightChar] !== undefined && charIndex[rightChar] >= left) {
            left = charIndex[rightChar] + 1;
        }
        
        charIndex[rightChar] = right;
        maxLength = Math.max(maxLength, right - left + 1);
    }
    
    return maxLength;
};

// More concise version
var lengthOfLongestSubstringConcise = function(s) {
    const seen = new Set();
    let left = 0, maxLength = 0;
    
    for (let right = 0; right < s.length; right++) {
        while (seen.has(s[right])) {
            seen.delete(s[left++]);
        }
        seen.add(s[right]);
        maxLength = Math.max(maxLength, right - left + 1);
    }
    
    return maxLength;
};

File: neetcode-150/sliding-window/longest-substring-without-repeating-characters/solution.py
"""
Time Complexity: O(n) - Each character visited at most twice
Space Complexity: O(min(m,n)) - Where m is charset size
"""
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        
        char_count = {}
        left = 0
        max_length = 0
        
        for right in range(len(s)):
            right_char = s[right]
            char_count[right_char] = char_count.get(right_char, 0) + 1
            
            # Contract window if duplicates found
            while char_count[right_char] > 1:
                left_char = s[left]
                char_count[left_char] -= 1
                left += 1
            
            max_length = max(max_length, right - left + 1)
        
        return max_length

# Alternative approach using Set
class SolutionSet:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        
        char_set = set()
        left = 0
        max_length = 0
        
        for right in range(len(s)):
            right_char = s[right]
            
            # Contract window if duplicate found
            while right_char in char_set:
                char_set.remove(s[left])
                left += 1
            
            char_set.add(right_char)
            max_length = max(max_length, right - left + 1)
        
        return max_length

# Optimized version with character position tracking
class SolutionOptimized:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_index = {}
        left = 0
        max_length = 0
        
        for right in range(len(s)):
            right_char = s[right]
            
            if right_char in char_index and char_index[right_char] >= left:
                left = char_index[right_char] + 1
            
            char_index[right_char] = right
            max_length = max(max_length, right - left + 1)
        
        return max_length

# More concise version
class SolutionConcise:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        left = max_length = 0
        
        for right in range(len(s)):
            while s[right] in seen:
                seen.remove(s[left])
                left += 1
            seen.add(s[right])
            max_length = max(max_length, right - left + 1)
        
        return max_length

# Using defaultdict for cleaner code
from collections import defaultdict

class SolutionDefaultDict:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_count = defaultdict(int)
        left = max_length = 0
        
        for right in range(len(s)):
            char_count[s[right]] += 1
            
            while char_count[s[right]] > 1:
                char_count[s[left]] -= 1
                left += 1
            
            max_length = max(max_length, right - left + 1)
        
        return max_length
